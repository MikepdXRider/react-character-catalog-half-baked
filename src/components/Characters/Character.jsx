export default function Character({ character }) {
  return (
    <figure>
      <img src={character.image} alt={character.name} />
      <figcaption>
        <h3>{character.name}</h3>
        <p>{character.species}</p>
        <p>Status: {character.status}</p>
      </figcaption>
    </figure>
  );
}
