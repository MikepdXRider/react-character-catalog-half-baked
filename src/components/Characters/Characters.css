.list {
  display: grid;
  margin: 0;
  padding: 0;
  gap: 1rem;
  margin-top: 2rem;
}

@media (min-width: 700px) {
  .list {
    grid: 1fr / 1fr 1fr 1fr 1fr;
  }
}

@media (max-width: 700px) {
  .list {
    grid: 1fr / 1fr 1fr;
  }
}

.item {
  list-style-type: none;
  text-align: center;
  padding: 5px;
  background: var(--light-grey);
  border-radius: 5px;
}

.image {
  max-width: 100%;
  border-radius: 5px;
  margin-bottom: 7px;
  box-shadow: 2px 2px 4px 2px rgba(0, 0, 0, 0.55);
}

.linkText {
  font-weight: bold;
  font-size: large;
  text-transform: uppercase;
}
