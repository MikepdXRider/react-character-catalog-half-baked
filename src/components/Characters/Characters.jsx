import { Link } from 'react-router-dom';
import styles from './Characters.css';

export default function Characters({ characters }) {
  return (
    <ul aria-label="characters" className={styles.list}>
      {characters.map((character) => (
        <li key={character.name} className={styles.item}>
          <Link to={`/characters/${character.id}`}>
            <figure>
              <img
                src={character.image}
                alt={character.name}
                className={styles.image}
              />
              <figcaption className={styles.linkText}>
                {character.name}
              </figcaption>
            </figure>
          </Link>
        </li>
      ))}
    </ul>
  );
}
