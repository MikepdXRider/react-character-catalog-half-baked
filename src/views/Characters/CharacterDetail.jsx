import { useEffect, useState } from 'react';
import { Link, useParams } from 'react-router-dom';
import Character from '../../components/Characters/Character';
import { getCharacter } from '../../services/rickAndMortyApi';

export default function CharacterDetail() {
  const { id } = useParams();
  const [loading, setLoading] = useState(true);
  const [character, setCharacter] = useState(null);

  useEffect(() => {
    getCharacter(id)
      .then(setCharacter)
      .finally(() => setLoading(false));
  }, [id]);

  return (
    <>
      <Link to="/characters">Back to Characters</Link>
      <article>
        {loading ? (
          <h1>Loading character...</h1>
        ) : (
          <Character character={character} />
        )}
      </article>
    </>
  );
}
