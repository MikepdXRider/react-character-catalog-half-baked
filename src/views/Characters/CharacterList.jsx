import { useEffect, useState } from 'react';
import Characters from '../../components/Characters/Characters';
import Pagination from '../../components/Pagination/Pagination';
import { usePagination } from '../../hooks/usePagination';
import { getCharacters } from '../../services/rickAndMortyApi';

import loadingSpinner from '../../assets/images/loading-spinner.gif';

export default function CharacterList() {
  const [characters, setCharacters] = useState([]);
  const [loading, setLoading] = useState(true);
  const {
    currentPage,
    totalPages,
    setCurrentPage,
    setTotalPages,
    viewNextPage,
    viewPrevPage,
  } = usePagination();

  useEffect(() => {
    setLoading(true);
    getCharacters(currentPage)
      .then((res) => {
        setTotalPages(res.pages);
        setCharacters(res.characters);
      })
      .finally(() => setLoading(false));
  }, [currentPage]);

  return (
    <>
      <h1>
        Characters{' '}
        {loading && (
          <img
            src={loadingSpinner}
            alt="loading characters"
            style={{ width: '20px', height: '20px', display: 'inline' }}
          />
        )}
      </h1>

      <Pagination
        currentPage={currentPage}
        totalPages={totalPages}
        setCurrentPage={setCurrentPage}
        viewNextPage={viewNextPage}
        viewPrevPage={viewPrevPage}
      />
      {!loading && <Characters characters={characters} />}
    </>
  );
}
