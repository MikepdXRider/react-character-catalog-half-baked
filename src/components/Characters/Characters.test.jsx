import {render, screen} from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom';
import Characters from './Characters.jsx'


it('characters details snapshot', async () => {
    const { container } = render(
        <MemoryRouter>
            <Characters 
            characters={[{
                image: 'test-image',
                name: 'test-name',
                species: 'test-species',
                status: 'test-status'
            }]}
            />
        </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
})