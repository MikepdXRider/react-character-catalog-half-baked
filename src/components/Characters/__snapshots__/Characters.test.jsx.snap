// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`characters details snapshot 1`] = `
<div>
  <ul
    aria-label="characters"
    class="list"
  >
    <li
      class="item"
    >
      <a
        href="/characters/undefined"
      >
        <figure>
          <img
            alt="test-name"
            class="image"
            src="test-image"
          />
          <figcaption
            class="linkText"
          >
            test-name
          </figcaption>
        </figure>
      </a>
    </li>
  </ul>
</div>
`;
